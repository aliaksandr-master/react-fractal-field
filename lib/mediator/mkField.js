export const mkMeta = () => ({
  valid: true,
  touched: false,

  active: false,
  activated: false,

  submitted: false,
  submitting: false,
  submitFailed: false,
  submitSuccess: false,
  submittedTimes: 0,

  error: null,
  errorsIn: [],
  hasException: false,
  hasFormatException: false,
  hasNormalizeException: false,
  hasValidateException: false
});

export default ({
  id,
  complexValueType = null,
  parentID = null,
  autoClean = false,
  shareMeta = true,
  shareValue = true,
  value = undefined,
  name = null
}) => ({
  // permanent
  id,
  parentID,
  autoClean,
  shareMeta,
  shareValue,
  complexValueType, // if not null component might have named children

  // free transform
  name,

  value,

  meta: mkMeta(),

  children: [],
  version: 0,
  metaVersion: 0,
  valueVersion: 0
});
