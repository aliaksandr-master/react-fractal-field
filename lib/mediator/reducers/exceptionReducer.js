import recalcValidity from '../recalcValidity';


export default (propName, exceptionPropsMap) => (state, { fieldID, payload: exception }) => {
  const field = state.fields[fieldID];

  exception = Boolean(exception);

  if (field[propName] === exception) {
    return state;
  }

  const hasException = exception || exceptionPropsMap[propName].some((propName) => Boolean(field[propName]));

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        valid: field.valid && !exception,
        hasException,
        [propName]: exception
      }
    }
  };

  return recalcValidity(state, field.parentID, field.id);
};
