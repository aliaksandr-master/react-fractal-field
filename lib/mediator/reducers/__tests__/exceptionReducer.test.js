/*eslint-env jest*/
import exceptionReducer from '../exceptionReducer';
import mkField from '../../mkField';


describe('exceptionReducer', () => {
  it('hasFormatException', () => {
    expect(exceptionReducer('hasFormatException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();

    expect(exceptionReducer('hasFormatException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: null })).toMatchSnapshot();

    expect(exceptionReducer('hasFormatException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: '' })).toMatchSnapshot();

    expect(exceptionReducer('hasFormatException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();
  });
  it('hasNormalizeException', () => {
    expect(exceptionReducer('hasNormalizeException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();

    expect(exceptionReducer('hasNormalizeException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: null })).toMatchSnapshot();

    expect(exceptionReducer('hasNormalizeException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: '' })).toMatchSnapshot();

    expect(exceptionReducer('hasNormalizeException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();
  });
  it('hasValidateException', () => {
    expect(exceptionReducer('hasValidateException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();

    expect(exceptionReducer('hasValidateException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: null })).toMatchSnapshot();

    expect(exceptionReducer('hasValidateException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: '' })).toMatchSnapshot();

    expect(exceptionReducer('hasValidateException')({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();
  });
});
