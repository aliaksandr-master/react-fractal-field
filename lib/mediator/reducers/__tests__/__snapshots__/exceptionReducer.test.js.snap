// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exceptionReducer hasFormatException 1`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": true,
      "hasFormatException": true,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 1,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 1,
  },
}
`;

exports[`exceptionReducer hasFormatException 2`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 0,
  },
}
`;

exports[`exceptionReducer hasFormatException 3`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 0,
  },
}
`;

exports[`exceptionReducer hasFormatException 4`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": true,
      "hasFormatException": true,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 1,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 1,
  },
}
`;

exports[`exceptionReducer hasNormalizeException 1`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": true,
      "hasFormatException": false,
      "hasNormalizeException": true,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 1,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 1,
  },
}
`;

exports[`exceptionReducer hasNormalizeException 2`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 0,
  },
}
`;

exports[`exceptionReducer hasNormalizeException 3`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 0,
  },
}
`;

exports[`exceptionReducer hasNormalizeException 4`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": true,
      "hasFormatException": false,
      "hasNormalizeException": true,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 1,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 1,
  },
}
`;

exports[`exceptionReducer hasValidateException 1`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": true,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": true,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 1,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 1,
  },
}
`;

exports[`exceptionReducer hasValidateException 2`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 0,
  },
}
`;

exports[`exceptionReducer hasValidateException 3`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 0,
  },
}
`;

exports[`exceptionReducer hasValidateException 4`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": true,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": true,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 1,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 1,
  },
}
`;
